packages/components/src/lib/controllers/ui-controller.js

```diff
@@ -57,39 +82,77 @@ export default class UIController {
 
     this._api = api
     this._store = store
-    this.uiOptions = uiOptions
+    this._uiOptions = uiOptions
+
+    // region Public API of a UI controller
+    // A public API must be defined before modules are created because modules may use it
+    this._api.ui.openPanel = this.openPanel.bind(this)
+    this._api.ui.closePanel = this.closePanel.bind(this)
+    this._api.ui.showPanelTab = this.showPanelTab.bind(this)
+    this._api.ui.changeTab = this.changeTab.bind(this)
+    this._api.ui.togglePanelTab = this.togglePanelTab.bind(this)
+    this._api.ui.openPopup = this.openPopup.bind(this)
+    this._api.ui.closePopup = this.closePopup.bind(this)
+    this._api.ui.isPopupVisible = () => Boolean(this.hasModule('popup') && this._store.state.popup.visible)
+    this._api.ui.openToolbar = this.openToolbar.bind(this)
+    this._api.ui.openActionPanel = this.openActionPanel.bind(this)
+    this._api.ui.closeActionPanel = this.closeActionPanel.bind(this)
+    this._api.ui.toggleActionPanel = this.toggleActionPanel.bind(this)
+    // endregion Public API of a UI controller
 
     // Set options of _modules before _modules are created
     if (this.hasModule('popup')) {
       let popupOptions = this._modules.get('popup').options // eslint-disable-line prefer-const
       popupOptions.initialShift = {
-        x: this.uiOptions.items.popupShiftX.currentValue,
-        y: this.uiOptions.items.popupShiftY.currentValue
+        x: this._uiOptions.items.popupShiftX.currentValue,
+        y: this._uiOptions.items.popupShiftY.currentValue
       }
     }
 
     if (this.hasModule('toolbar')) {
       let toolbarOptions = this._modules.get('toolbar').options // eslint-disable-line prefer-const
       toolbarOptions.initialShift = {
-        x: this.uiOptions.items.toolbarShiftX.currentValue,
-        y: this.uiOptions.items.toolbarShiftY.currentValue
+        x: this._uiOptions.items.toolbarShiftX.currentValue,
+        y: this._uiOptions.items.toolbarShiftY.currentValue
       }
     }
 
     this.createModules()
 
     // Adjust configuration of _modules according to feature options
     if (this.hasModule('panel')) {
-      this._store.commit('panel/setPosition', this.uiOptions.items.panelPosition.currentValue)
+      this._store.commit('panel/setPosition', this._uiOptions.items.panelPosition.currentValue)
     }
+
+    this.setFontSize(this._uiOptions)
   }
 
   activate () {
+    this._uiState.activate()
     this.activateModules()
+    this._uiState.activateUI()
+
+    if (this.hasModule('panel')) {
+      if (this._uiState.isPanelStateDefault() || !this._uiState.isPanelStateValid()) {
+        this._uiState.setPanelClosed()
+      }
+      if (this._uiState.isPanelOpen()) {
+        this.openPanel(true)
+      }
```

Comment:
this logic is a little different than before. is there any chance the panel will be left in an undefined state?

Reply:
I thought it is an equivalent but I might be missing something.
In master, it is:
```
    if (this.state.isPanelStateDefault() || !this.state.isPanelStateValid()) {
      this.setDefaultPanelState()
    }
    // If panel should be opened according to the state, open it
    if (this.state.isPanelOpen()) {
      if (this.uic.hasModule('panel')) { this.api.ui.openPanel(true) } // Force close the panel
    }
```
(see https://github.com/alpheios-project/alpheios-core/blob/master/packages/components/src/lib/controllers/app-controller.js#L997-L1003)

and as `setDefaultPanelState()` is defined as
```
  setDefaultPanelState () {
    if (!this.uic.hasModule('panel')) { return this }
    this.state.setPanelClosed()
    return this
  }
```
Then, combined, it will be:
```
    if (this.state.isPanelStateDefault() || !this.state.isPanelStateValid()) {
      if (!this.uic.hasModule('panel')) { return this } // I.e. do nothing
      this.state.setPanelClosed()
    }
    // If panel should be opened according to the state, open it
    if (this.state.isPanelOpen()) {
      if (this.uic.hasModule('panel')) { this.api.ui.openPanel(true) } // Force close the panel
    }
```
In both cases nothing is done if there is no `panel` module available. We open a panel if its state is `open` and close if the state is either `default` or invalid. In all other cases we do nothing. So I think it must be pretty much what we had before.

The complicated part is that the `state` is provided from an outside of an app: both webextension and embed lib have their own state object. That bothers me because an app controller cannot know what possible states will it be dealing with. Because of this we, in the app controller, cannot expect anything about the states being provided. I think this is no good.

That's one of the reasons I was suggesting to simplify state management. I as thinking about using Vuex but I agree it's not a very good idea because of a dependency that it will create. But I still think something can and needs to be done here. Maybe we can do the change when we'll convert to code to adapt to the Manifest V3, because it's a webextension that using the state the most?

What do you think about that? Can you suggest any improvements to the current piece of code?

Reply:
I think the current code is fine for now.

I agree that we need to deal with the co-managed state (client app vs app controller) better. Not sure when or how is best to do that right not, but for the moment waiting until we adapt for Manifest 3 seems as good of an idea as any!